packages/jupyter-ai/jupyter_ai/config_manager.py

```diff
@@ -12,120 +16,317 @@
+        config file conforms to the JSON Schema, and that the language and
+        embedding models have authn credentials if specified."""
+        self.validator.validate(config.dict())
+
+        # validate language model config
+        if config.model_provider_id:
+            _, lm_provider = get_lm_provider(
+                config.model_provider_id, self._lm_providers
+            )
+            if not lm_provider:
                 raise ValueError(
-                    f"Missing value for '{auth_strategy.name}' in the config."
+                    f"No language model is associated with '{config.model_provider_id}'."
                 )
-            provider_params[name.lower()] = api_keys[name]
+            _validate_provider_authn(config, lm_provider)
 
-    def _update_lm_provider(self, config: GlobalConfig):
-        model_id = config.model_provider_id
+        # validate embedding model config
+        if config.embeddings_provider_id:
+            _, em_provider = get_em_provider(
+                config.embeddings_provider_id, self._em_providers
+            )
+            if not em_provider:
+                raise ValueError(
+                    f"No embedding model is associated with '{config.embeddings_provider_id}'."
+                )
+            _validate_provider_authn(config, em_provider)
 
-        if not model_id:
-            self.lm_provider = None
-            self.lm_provider_params = None
-            return
+    def _write_config(self, new_config: GlobalConfig):
+        """Updates configuration and persists it to disk. This accepts a
+        complete `GlobalConfig` object, and should not be called publicly."""
+        # remove any empty field dictionaries
+        new_config.fields = {k: v for k, v in new_config.fields.items() if v}
 
-        local_model_id, provider = get_lm_provider(model_id, self.lm_providers)
+        self._validate_config(new_config)
+        with open(self.config_path, "w") as f:
+            json.dump(new_config.dict(), f, indent=self.indentation_depth)
 
-        if not provider:
-            raise ValueError(f"No provider and model found with '{model_id}'")
+    def delete_api_key(self, key_name: str):
+        config_dict = self._read_config().dict()
+        lm_provider = self.lm_provider
+        em_provider = self.em_provider
+        required_keys = []
+        if (
+            lm_provider
+            and lm_provider.auth_strategy
+            and lm_provider.auth_strategy.type == "env"
+        ):
+            required_keys.append(lm_provider.auth_strategy.name)
+        if (
+            em_provider
+            and em_provider.auth_strategy
+            and em_provider.auth_strategy.type == "env"
+        ):
+            required_keys.append(self.em_provider.auth_strategy.name)
 
-        fields = config.fields.get(model_id, {})
-        provider_params = {"model_id": local_model_id, **fields}
+        if key_name in required_keys:
+            raise KeyInUseError(
```

Comment:
Could we pop up a confirmation dialog instead, that would both delete the API key and reset the language/embedding model? I'm concerned that the `raise` here might introduce some excessive friction when a user wants to unset an API key.

Reply:
This is a backend file. We have to perform this validation, or else the currently selected providers will break. This isn't excessive friction.

If this error is confusing, we can add client side validation to simply disable the "delete" button when the key is being used by the currently selected providers.